kyo-prelude/shared/src/test/scala/kyo2/ChoiceTest.scala

```diff
@@ -0,0 +1,98 @@
+package kyo2
+
+class ChoiceTest extends Test:
+
+    "eval with a single choice" in {
+        assert(
+            kyo2.Choice.run(kyo2.Choice.eval(Seq(1))(i => (i + 1))).eval == Seq(2)
+        )
+    }
+
+    "eval with multiple choices" in {
+        assert(
+            kyo2.Choice.run(kyo2.Choice.eval(Seq(1, 2, 3))(i => (i + 1))).eval == Seq(2, 3, 4)
+        )
+    }
+
+    "nested eval" in {
+        assert(
+            kyo2.Choice.run(kyo2.Choice.eval(Seq(1, 2, 3))(i =>
+                kyo2.Choice.get(Seq(i * 10, i * 100))
+            )).eval == Seq(10, 100, 20, 200, 30, 300)
+        )
+    }
+
+    "drop" in {
+        assert(
+            kyo2.Choice.run(kyo2.Choice.eval(Seq(1, 2, 3))(i =>
+                if i < 2 then kyo2.Choice.drop else kyo2.Choice.get(Seq(i * 10, i * 100))
+            )).eval == Seq(20, 200, 30, 300)
+        )
+    }
+
+    "filter" in {
+        assert(
+            kyo2.Choice.run(kyo2.Choice.eval(Seq(1, 2, 3))(i =>
+                kyo2.Choice.dropIf(i < 2).map(_ => kyo2.Choice.get(Seq(i * 10, i * 100)))
+            )).eval == Seq(20, 200, 30, 300)
+        )
+    }
+
+    "empty choices" in {
+        assert(
+            kyo2.Choice.run(kyo2.Choice.eval(Seq.empty[Int])(_ => 42)).eval == Seq.empty[Int]
+        )
+    }
+
+    "nested drop" in {
+        assert(
+            kyo2.Choice.run(
+                kyo2.Choice.eval(Seq(1, 2, 3))(i =>
+                    kyo2.Choice.eval(Seq(i * 10, i * 100))(j =>
+                        if j > 100 then kyo2.Choice.drop else j
+                    )
+                )
+            ).eval == Seq(10, 100, 20, 30)
+        )
+    }
+
+    "nested filter" in {
+        assert(
+            kyo2.Choice.run(
+                kyo2.Choice.eval(Seq(1, 2, 3))(i =>
+                    kyo2.Choice.dropIf(i % 2 != 0).map(_ =>
+                        kyo2.Choice.eval(Seq(i * 10, i * 100))(j =>
+                            kyo2.Choice.dropIf(j >= 300).map(_ => j)
+                        )
+                    )
+                )
+            ).eval == Seq(20, 200)
+        )
+    }
+
+    "large number of choices" in {
+        val largeChoice = Seq.range(0, 100000)
+        try
+            assert(
+                kyo2.Choice.run(kyo2.Choice.get(largeChoice)).eval == largeChoice
+            )
+        catch
+            case ex: StackOverflowError => fail()
+        end try
+    }
+
+    "large number of suspensions" in pendingUntilFixed {
```

Comment:
This is still stack-unsafe because the nested `Choice.run` call during effect handling doesn't go through the new stack depth tracking mechanism. We can probably fix it later.